### ISSUE TITLE
Block all call forwarding

### ISSUE THREAD

**Issue by CyanDragonS** [aj3423/SpamBlocker#155]
Many spam calls try to use call forwarding to bypass spam blocker, if there can be a function to block call forwarding incoming calls?

**Comment by aj3423** [comment]
While STIR/SHAKEN can provide some information about the caller's identity, it cannot definitively determine whether a call is forwarded. 

> Many spam calls try to use call forwarding to bypass spam blocker

How does that work? And how can you tell if a call is forwarded?

**Comment by CyanDragonS** [comment]
At the moment, I don't really know how to use Android's internal api to determine whether a call is a forwarding call. But when a call comes in, my phone call screen will show that it is a call forwarding call. My phone is a MIUI system, which comes with a system-level spam block that can identify call transfers and block them, but my other phone can't.

**Comment by aj3423** [comment]
Interesting, is there a built-in call screening feature in MIUI?

> but my other phone can't.

I've seen Samsung has similar feature, it doesn't show "forwarded" information, it shows the peer carrier name, like: "XXX Telecom". I analyzed that Samsung Dialer app and there is a built-in database, maybe MIUI also has such database for call screening. I guess if you replace the dialer with any other app, there will not be such infomation any more.

**Comment by aj3423** [comment]
I guess it's a dialer thing, I don't find anything releated to "forwarding" in the [doc](https://developer.android.com/reference/android/telecom/Call.Details), I'll look into this later.

**Comment by aj3423** [comment]
Didn't find anything useful, google shows all kinds of irrelevant stuff. Can you show me the package name and the version number of that dialer app? I'll try to analyze that apk.

The package name can be found in the Recent Apps list.
The version number can be found in the app settings.

**Comment by aj3423** [comment]
The securitycenter has nothing to do with this.
The incallui is only used for displaying the UI during the call, this is how it checks if it `isForwardedCall()`:
```java
class PhoneConstants {
	public static final String EXTRA_MIUI_CALL_EXTRAS = "com.miui.phone.CALL_EXTRAS";
	public static final String EXTRA_IS_FORWARDED_CALL = "telephony.IS_FORWARDED_CALL";    <---- this 
}

class CallAdapterUtils {

	public static Bundle getCallExtra(android.telecom.Call call) {
		Bundle extras;
		if (call == null || (extras = call.getDetails().getExtras()) == null) {
			return null;
		}
		return extras.getBundle(PhoneConstants.EXTRA_MIUI_CALL_EXTRAS);
	}

}

class Call {
	private android.telecom.Call mTelecommCall;

	public Bundle getCallExtra() {
		return CallAdapterUtils.getCallExtra(this.mTelecommCall);
	}

	public boolean isForwardedCall() {     <-------- this
		Bundle callExtra = getCallExtra();
		return callExtra != null && callExtra.getBoolean(PhoneConstants.EXTRA_IS_FORWARDED_CALL, false);
	}
}
```

The key seems to be the "telephony.IS_FORWARDED_CALL" in the intent, it's not set by this app itself, I think this intent is sent from the Phone app, what is the package name and version code of the Phone app?

**Comment by aj3423** [comment]
Interesting, calls are handled by a contact management app.

I won't be adding feature for any particular brand, so if it relies on its own database, it won't be supported, even if they provide API.
But if there's some flag in the call information, it can be a very useful feature.
I'll try to analyze that app.

**Comment by aj3423** [comment]
> Some apps can operate to this 'android' blacklist database

It's the android built-in [BlockNumberContract](https://developer.android.com/reference/android/provider/BlockedNumberContract#overview)

> Only the system, the default SMS application, and the default phone app , and carrier apps can read, and write to the blockednumber provider.

The call won't even reach the call screening app when the number exists in the BlockNumberContract.

I've been trying to trace this `com.miui.phone.CALL_EXTRAS`, which is an attribute sent from the MIUI framework, but I don't find any clue in the contacts or security app, can't go any further.

Anyway, it seems like a MIUI specific thing, I'm not gonna add feature for any particular brand.

**Comment by aj3423** [comment]
Maybe the number is checked on the fly, it's sent to the cloud server. This can be verified by turning off both Wi-Fi and mobile data and see if it still shows the "forwarded call".

**Comment by aj3423** [comment]
> See PLS this link:

Just learned from this link, Samsung do have the "forwarded" indicator, it's a small toast message that will last for only 1 second, very easy to miss.

Now I agree that it's an system level attribute.

- For Samsung, the toast is shown by the com.android.phone, this is an system level app that only has odex. There is no "forwarded" information in the CallScreeningService at all. The attached file contains the detaild analyzing steps, for future reference.
- For MIUI, the system passes this infomation to the CallDetails, as the attribute "telephony.IS_FORWARDED_CALL". Not verified, though, I don't have a MIUI device.

To summarize, this "forwarded" information is available in the ROM level, some manufacture will forward it to the phone app, but some will ignore it completely. 

On second thought, this feature it not that useful, even less useful than the due sim support. Anyway,  who wants this should fire an feature request to the Android devs, asking for forwarding this attribute to the CallScreeningService, so as to make it a standard.

[trace.log](https://github.com/user-attachments/files/17495478/trace.log)
[TeleService.dex.zip](https://github.com/user-attachments/files/17495483/TeleService.dex.zip)
